lib/TPP/Conversion/ConvertLinalgToXsmm/ConvertLinalgToXsmm.cpp

```diff
@@ -323,66 +323,90 @@ getBroadCastBinaryFlagFromMap(AffineMap map, unsigned operandIdx) {
   }
 }
 
-// Convert linalg.generic to xsmm binary add op.
-struct ConvertGenericToBinaryAdd : public OpRewritePattern<linalg::GenericOp> {
+static LogicalResult rewriteBinaryOp(RewriterBase &rewriter,
+                                     linalg::GenericOp genericOp,
+                                     ArrayRef<Value> operands,
+                                     xsmm::BinaryKind xsmmTy) {
+  assert(operands.size() == 3);
+  auto lhs = operands[0];
+  auto rhs = operands[1];
+  auto output = operands[2];
+
+  OpOperand *lhsOperand = getOperandFromValue(genericOp, lhs);
+  auto broadCastFlagLhs = getBroadCastBinaryFlagFromMap(
+      genericOp.getMatchingIndexingMap(lhsOperand), /*operandIdx=*/0);
+  if (failed(broadCastFlagLhs))
+    return failure();
+
+  OpOperand *rhsOperand = getOperandFromValue(genericOp, rhs);
+  auto broadCastFlagRhs = getBroadCastBinaryFlagFromMap(
+      genericOp.getMatchingIndexingMap(rhsOperand), /*operandIdx=*/1);
+  if (failed(broadCastFlagRhs))
+    return failure();
+
+  auto binaryInfo = xsmm::utils::getBinaryInfo(lhs, *broadCastFlagLhs, rhs,
+                                               *broadCastFlagRhs, output);
+  if (failed(binaryInfo))
+    return failure();
+
+  auto flagLhs =
+      xsmm::BinaryFlagsAttr::get(rewriter.getContext(), *broadCastFlagLhs);
+  auto flagRhs =
+      xsmm::BinaryFlagsAttr::get(rewriter.getContext(), *broadCastFlagRhs);
+
+  // Spaghetti code to handle 'NONE' as it conflicts with other flags; we
+  // cannot add it if at least the RHS or the LHS is not 'NONE'. Maybe the
+  // best solution is to get rid of it.
+  ArrayAttr flags;
```

Comment:
For a different PR...

Perhaps this could go to a utility function somewhere else. Looks fairly independent.

Though `NONE` is just `0`, you can `if (!flagsLhs.getValue())`. Or you could make the attribute optional and not create it in the first place (for NONE)?

Also, you could just create a small vector and create the array attr from the vector?
```c++
  SmallVector flags;
  if (flagsLhs.getValue())
    flags.push_back(flagsLhs.getValue());
  if (flagsRhs.getValue())
    flags.push_back(flagsRhs.getValue());
  ArrayAttr flagsAttr = rewriter.getArrayAttr(flags);
```

Reply:
`getValue()` return the attribute value and it is not convertible to zero (you still need to have `!= xsmm::BinaryFlags::NONE`). Switch the rest to use a SmallVector

Reply:
The NONE business is a know annoying problem. I should have made it optional but at the moment is not (you cannot have an empty flags vector).

Reply:
No worries, it can be another PR.